### PR TITLE
write once delete once init

### PR DESCRIPTION
**Pull request by Sam-Kramer**
## General
**Before this PR**:

**After this PR**:

==COMMIT_MSG==
Adds write-once-delete-once workflow
==COMMIT_MSG==

**Priority**:

**Concerns / possible downsides (what feedback would you like?)**:

**Is documentation needed?**:

## Compatibility
**Does this PR create any API breaks (e.g. at the Java or HTTP layers) - if so, do we have compatibility?**:

**Does this PR change the persisted format of any data - if so, do we have forward and backward compatibility?**:

**The code in this PR may be part of a blue-green deploy. Can upgrades from previous versions safely coexist? (Consider restarts of blue or green nodes.)**:

**Does this PR rely on statements being true about other products at a deployment - if so, do we have correct product dependencies on these products (or other ways of verifying that these statements are true)?**:

**Does this PR need a schema migration?**

## Testing and Correctness
**What, if any, assumptions are made about the current state of the world? If they change over time, how will we find out?**:

**What was existing testing like? What have you done to improve it?**:

**If this PR contains complex concurrent or asynchronous code, is it correct? The onus is on the PR writer to demonstrate this.**:

**If this PR involves acquiring locks or other shared resources, how do we ensure that these are always released?**:

## Execution
**How would I tell this PR works in production? (Metrics, logs, etc.)**:

**Has the safety of all log arguments been decided correctly?**:

**Will this change significantly affect our spending on metrics or logs?**:

**How would I tell that this PR does not work in production? (monitors, etc.)**:

**If this PR does not work as expected, how do I fix that state? Would rollback be straightforward?**:

**If the above plan is more complex than “recall and rollback”, please tag the support PoC here (if it is the end of the week, tag both the current and next PoC)**:

## Scale
**Would this PR be expected to pose a risk at scale? Think of the shopping product at our largest stack.**:

**Would this PR be expected to perform a large number of database calls, and/or expensive database calls (e.g., row range scans, concurrent CAS)?**:

**Would this PR ever, with time and scale, become the wrong thing to do - and if so, how would we know that we need to do something differently?**:

## Development Process
**Where should we start reviewing?**:

**If this PR is in excess of 500 lines excluding versions lock-files, why does it not make sense to split it?**:

**Please tag any other people who should be aware of this PR**:
@jeremyk-91
@sverma30
@raiju

### CODE REVIEW

File: atlasdb-workload-server-antithesis/var/compact.sh [hunk 1]
```diff
@@ -16,4 +16,4 @@
 # limitations under the License.
 #
 
-podman exec $1 bash -c 'palantir-cassandra-*/bin/nodetool compact'¡
\ No newline at end of file
+podman exec $1 bash -c 'palantir-cassandra-*/bin/nodetool compact'
```
Comment (Sam-Kramer) [inline atlasdb-workload-server-antithesis/var/compact.sh:19]: Includes fixes for our compact/flush scripts!

Comment (Sam-Kramer) [reply atlasdb-workload-server-antithesis/var/compact.sh:19]: As this workflow revealed they weren't working.

File: atlasdb-workload-server-antithesis/var/docker-compose.yml [hunk 2]
```diff
@@ -20,7 +20,7 @@ services:
         ipv4_address: 10.20.20.2
 
   cassandra2:
-    image: palantirtechnologies/cassandra:2.2.18-1.112.0-rc5
+    image: palantirtechnologies/cassandra:2.2.18-1.112.0-rc8
```
Comment (Sam-Kramer) [inline atlasdb-workload-server-antithesis/var/docker-compose.yml:23]: And bumps to the latest C* RC, which are nice to have but not critical for this PR. Happy to split if reviewer deems necessary.

File: atlasdb-workload-server-distribution/src/main/java/com/palantir/atlasdb/workload/WorkloadServerLauncher.java [hunk 3]
```diff
@@ -119,12 +121,11 @@ private void scheduleBackgroundJobs(Environment environment) {
                         AntithesisCassandraSidecarResource.INSTANCE,
                         DefaultBuggifyFactory.INSTANCE),
                 0,
-                10,
+                2,
```
Comment (Sam-Kramer) [inline atlasdb-workload-server-distribution/src/main/java/com/palantir/atlasdb/workload/WorkloadServerLauncher.java:124]: We need to compact/flush _way_ more often

File: atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/workflow/WriteOnceDeleteOnceWorkflows.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        public Optional<WitnessedTransaction> run(InteractiveTransactionStore store, int taskIndex) {
+            configuration.transactionRateLimiter().acquire();
+            return store.readWrite(txn -> {
+                Integer rowKey = rowKeyGenerator.apply(configuration.maxKey());
+                if (buggifyFactory
+                        .maybe(calculateChanceForDelete(taskIndex, configuration.iterationCount()))
+                        .asBoolean()) {
+                    txn.delete(configuration.tableConfiguration().tableName(), cell(rowKey));
+                } else {
+                    txn.write(configuration.tableConfiguration().tableName(), cell(rowKey), RANDOM.nextInt(10000));
```
Comment (Sam-Kramer) [inline atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/workflow/WriteOnceDeleteOnceWorkflows.java:90]: Using `RANDOM` here as we really never need this to be deterministic for testing purposes. This also makes usages of `rowKeyGenerator` for tests a lot easier to reason about / think.

File: atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/workflow/WriteOnceDeleteOnceWorkflowTest.java [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+    public void calculateChanceForDeleteHasExpectedValues() {
+        assertThat(WriteOnceDeleteOnceWorkflows.calculateChanceForDelete(0, 100))
+                .isEqualTo(0.01);
+        assertThat(WriteOnceDeleteOnceWorkflows.calculateChanceForDelete(99, 100))
+                .isEqualTo(1.00);
+    }
+
+    @Test
+    public void writesThenDeletes() {
+        Iterator<Integer> keys = List.of(0, 0, 1, 1).iterator();
```
Comment (Sam-Kramer) [inline atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/workflow/WriteOnceDeleteOnceWorkflowTest.java:94]: Probably should just do `Iterators.cycle`, but at the same time this is a bit easier to understand.

File: atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/workflow/WriteOnceDeleteOnceWorkflowTest.java [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+        BuggifyFactory alternatingBuggify = mock(BuggifyFactory.class);
+        when(alternatingBuggify.maybe(anyDouble()))
+                .thenReturn(
+                        NoOpBuggify.INSTANCE, DefaultBuggify.INSTANCE, NoOpBuggify.INSTANCE, DefaultBuggify.INSTANCE);
+        Workflow workflow = createWorkflow(alternatingBuggify, rowKeyGenerator, 4);
+
+        List<WitnessedTransaction> transactions = workflow.run().history();
+
+        assertThat(transactions.size()).isEqualTo(4);
+        assertThat(getTransactionAction(transactions.get(0), WitnessedWriteTransactionAction.class)
```
Comment (Sam-Kramer) [inline atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/workflow/WriteOnceDeleteOnceWorkflowTest.java:105]: should just make `assertWriteTransaction` and `assertDeleteTransaction`

File: atlasdb-workload-server-distribution/src/main/java/com/palantir/atlasdb/workload/WorkloadServerLauncher.java [hunk 7]
```diff
@@ -119,12 +121,11 @@ private void scheduleBackgroundJobs(Environment environment) {
...
                 TimeUnit.SECONDS);
     }
 
     private void runWorkflows(WorkloadServerConfiguration configuration, Environment environment) {
-        // This is a single threaded executor; this is intentional, so that we only run one workflow at a time.
```
Comment (jeremyk-91) [inline atlasdb-workload-server-distribution/src/main/java/com/palantir/atlasdb/workload/WorkloadServerLauncher.java:127]: As far as I can tell this is still true, so we should probably leave the comment (if we don't want to "only run one workflow at a time" we can reword it to say we want to test with running workflows in parallel or something)

File: atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/workflow/WriteOnceDeleteOnceWorkflows.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    @VisibleForTesting
+    static Double calculateChanceForDelete(int taskIndex, int iterationCount) {
+        // Index generation is [inclusive, exclusive], so we need to add 1 to the task index.
+        return (taskIndex + 1.0) / iterationCount;
+    }
+
+    @VisibleForTesting
+    static WorkloadCell cell(Integer index) {
```
Comment (jeremyk-91) [inline atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/workflow/WriteOnceDeleteOnceWorkflows.java:103]: nit: we've used boxed and unboxed int across the methods, we should probably pick a convention and stick with it!

File: atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/workflow/WriteOnceDeleteOnceWorkflows.java [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+                WriteOnceDeleteOnceWorkflowConfiguration configuration) {
+            this.buggifyFactory = buggifyFactory;
+            this.rowKeyGenerator = rowKeyGenerator;
+            this.configuration = configuration;
+        }
+
+        public Optional<WitnessedTransaction> run(InteractiveTransactionStore store, int taskIndex) {
+            configuration.transactionRateLimiter().acquire();
+            return store.readWrite(txn -> {
+                Integer rowKey = rowKeyGenerator.apply(configuration.maxKey());
```
Comment (jeremyk-91) [inline atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/workflow/WriteOnceDeleteOnceWorkflows.java:84]: maybe not actionable: I think this works out fine in the end, but be aware that `maxKey` is never actually generated because of exclusivity so the name is potentially misleading

File: atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/workflow/WriteOnceDeleteOnceWorkflowTest.java [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+import com.palantir.atlasdb.workload.transaction.witnessed.WitnessedTransactionAction;
+import com.palantir.atlasdb.workload.transaction.witnessed.WitnessedWriteTransactionAction;
+import com.palantir.atlasdb.workload.util.AtlasDbUtils;
+import java.util.Iterator;
+import java.util.List;
+import java.util.function.IntFunction;
+import org.junit.Test;
+
+public class WriteOnceDeleteOnceWorkflowTest {
+    private static final String TABLE_NAME = "my.latte";
```
Comment (jeremyk-91) [inline atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/workflow/WriteOnceDeleteOnceWorkflowTest.java:49]: ☕

File: atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/workflow/WriteOnceDeleteOnceWorkflowTest.java [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+        List<WitnessedTransaction> transactions = workflow.run().history();
+
+        assertThat(transactions.size()).isEqualTo(4);
+        assertWriteTransaction(transactions.get(0), 0);
+        assertDeleteTransaction(transactions.get(1), 0);
+        assertWriteTransaction(transactions.get(2), 1);
+        assertDeleteTransaction(transactions.get(3), 1);
+    }
+
+    private void assertWriteTransaction(WitnessedTransaction transaction, int key) {
```
Comment (jeremyk-91) [inline atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/workflow/WriteOnceDeleteOnceWorkflowTest.java:111]: nit: some of these methods should probably be static?

File: atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/workflow/WriteOnceDeleteOnceWorkflowTest.java (hunk 5 above)
Comment (jeremyk-91) [reply atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/workflow/WriteOnceDeleteOnceWorkflowTest.java:94]: The other sequence is valid too right? (write 0, delete 1, delete 0, write 1)

File: atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/workflow/WriteOnceDeleteOnceWorkflowTest.java [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+                        NoOpBuggify.INSTANCE, DefaultBuggify.INSTANCE, NoOpBuggify.INSTANCE, DefaultBuggify.INSTANCE);
+        Workflow workflow = createWorkflow(alternatingBuggify, rowKeyGenerator, 4);
+
+        List<WitnessedTransaction> transactions = workflow.run().history();
+
+        assertThat(transactions.size()).isEqualTo(4);
+        assertWriteTransaction(transactions.get(0), 0);
+        assertDeleteTransaction(transactions.get(1), 0);
+        assertWriteTransaction(transactions.get(2), 1);
+        assertDeleteTransaction(transactions.get(3), 1);
```
Comment (jeremyk-91) [inline atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/workflow/WriteOnceDeleteOnceWorkflowTest.java:108]: not actionable: the orchestration feels a bit heavy-handed... but I don't really have a nicer solution.

File: atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/workflow/WriteOnceDeleteOnceWorkflows.java [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+import com.palantir.atlasdb.workload.transaction.witnessed.WitnessedTransaction;
+import java.security.SecureRandom;
+import java.util.Optional;
+import java.util.function.IntFunction;
+
+/**
+ * The idea of this class is to catch bugs around tombstone drops, by having a
+ * workflow which _mostly_ writes once then deletes once for a given cell. Probability and randomness are used to
+ * enforce this behaviour, thus it's possible to have a write without a delete, or visa versa; some cells may
+ * randomly be written or deleted multiple times as well.
```
Comment (jeremyk-91) [inline atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/workflow/WriteOnceDeleteOnceWorkflows.java:36]: do we want to be more specific (initial transactions _mostly_ write, later transactions _mostly_ delete)?

**Review by jeremyk-91 — COMMENTED** [review]
Looks good!

File: atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/workflow/WriteOnceDeleteOnceWorkflows.java (hunk 9 above)
Comment (jeremyk-91) [reply atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/workflow/WriteOnceDeleteOnceWorkflows.java:84]: probably a comment should suffice

File: atlasdb-workload-server-distribution/src/main/java/com/palantir/atlasdb/workload/WorkloadServerLauncher.java (hunk 7 above)
Comment (Sam-Kramer) [reply atlasdb-workload-server-distribution/src/main/java/com/palantir/atlasdb/workload/WorkloadServerLauncher.java:127]: yikes -- yeah I tried running multiple workflows in parallel, and that broke things :( will re-add the comment

File: atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/workflow/WriteOnceDeleteOnceWorkflows.java (hunk 9 above)
Comment (Sam-Kramer) [reply atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/workflow/WriteOnceDeleteOnceWorkflows.java:84]: Added it to the configuration as a comment!

File: atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/workflow/WriteOnceDeleteOnceWorkflowTest.java (hunk 5 above)
Comment (Sam-Kramer) [reply atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/workflow/WriteOnceDeleteOnceWorkflowTest.java:94]: Yup of course, I was just modeling this to be closer to how it would be in practice, which would not be write-delete-write-delete

**Review by jeremyk-91 — APPROVED** [review]
👍 great, looks good!

**Review by ergo14 — APPROVED** [review]
👍🏼 rubber stamp as required
